pandas/core/frame.py

```diff
@@ -3007,6 +3007,12 @@ def _setitem_array(self, key, value):
                 for k1, k2 in zip(key, value.columns):
                     self[k1] = value[k2]
             else:
+                if all(is_hashable(k) for k in key):
```

Comment:
Under what cases can elements of `key` not be hashable? These are our eventual column names, right? Which I *think* we require they be hashable.

I'm a bit concerned about the performance here, for the case when `key` is large (which is possible, right?)

Reply:
The only reason that I check whether `key` is hashable is to produce the correct? errors, which is tested in https://github.com/pandas-dev/pandas/blob/e55b6980fd5c99087d0959952b22e74434e9f5e2/pandas/tests/indexing/test_indexing.py#L180-L184
If producing all these errors are not required then this check can be omitted.

Reply:
@TomAugspurger what's your thought on this?

Reply:
What happens if the check is removed?

Reply:
The error produced is different.

Reply:
Yea this seems strange to me. I feel like this *should* be hashable inherently so clarification here would be helpful

Reply:
Without testing `is_hashable`, at https://github.com/pandas-dev/pandas/blob/3c0cf22909804ae050fdf7609edd13d88d0460aa/pandas/tests/indexing/test_indexing.py#L184 may throw `Buffer has wrong number of dimensions (expected 1, got 2)` or `Array conditional must be same shape as self` instead of the expected errors.